### PR TITLE
exchange: stop, stop loss and take profit functions

### PR DESCRIPTION
**Pull request by Dan-krm**
Add some new methods to the base exchange class: `createTriggerOrder`, `createStopLossOrder`, `createTakeProfitOrder`, `createOrderWithTakeProfitAndStopLoss` and tested them on okx:

### createTriggerOrder:
```
okx createTriggerOrder BTC/USDT:USDT limit buy 1 50000 49000 '{"posSide":"long"}'

okx.createTriggerOrder (BTC/USDT:USDT, limit, buy, 1, 50000, 49000, [object Object])
2023-12-20T08:11:13.976Z iteration 0 passed in 369 ms

{
  info: {
    algoClOrdId: 'e847386590ce4dBC7eba5daba4bc902c',
    algoId: '657616010580852736',
    clOrdId: 'e847386590ce4dBC7eba5daba4bc902c',
    sCode: '0',
    sMsg: '',
    tag: 'e847386590ce4dBC'
  },
  id: '657616010580852736',
  clientOrderId: 'e847386590ce4dBC7eba5daba4bc902c',
  symbol: 'BTC/USDT:USDT',
  type: 'limit',
  side: 'buy',
  trades: [],
  reduceOnly: false,
  fees: []
}
```

### createTakeProfitOrder:
```
okx createTakeProfitOrder BTC/USDT:USDT limit sell 1 50000 49000 '{"posSide":"long"}'

okx.createTakeProfitOrder (BTC/USDT:USDT, limit, sell, 1, 50000, 49000, [object Object])
2023-12-20T08:17:13.046Z iteration 0 passed in 376 ms

{
  info: {
    algoClOrdId: 'e847386590ce4dBC8a51e4172402892a',
    algoId: '657617516633604096',
    clOrdId: 'e847386590ce4dBC8a51e4172402892a',
    sCode: '0',
    sMsg: '',
    tag: ''
  },
  id: '657617516633604096',
  clientOrderId: 'e847386590ce4dBC8a51e4172402892a',
  symbol: 'BTC/USDT:USDT',
  type: 'limit',
  side: 'sell',
  trades: [],
  reduceOnly: false,
  fees: []
}
```

### createStopLossOrder:
```
okx createStopLossOrder BTC/USDT:USDT limit sell 1 31000 30000 '{"posSide":"long"}'

okx.createStopLossOrder (BTC/USDT:USDT, limit, sell, 1, 31000, 30000, [object Object])
2023-12-20T08:19:44.018Z iteration 0 passed in 391 ms

{
  info: {
    algoClOrdId: 'e847386590ce4dBC062a009804dcb58e',
    algoId: '657618149818318848',
    clOrdId: 'e847386590ce4dBC062a009804dcb58e',
    sCode: '0',
    sMsg: '',
    tag: ''
  },
  id: '657618149818318848',
  clientOrderId: 'e847386590ce4dBC062a009804dcb58e',
  symbol: 'BTC/USDT:USDT',
  type: 'limit',
  side: 'sell',
  trades: [],
  reduceOnly: false,
  fees: []
}
```

### createOrderWithTakeProfitAndStopLoss:
```
okx createOrderWithTakeProfitAndStopLoss LTC/USDT:USDT limit buy 1 60 70 50 '{"posSide":"long"}'

okx.createOrderWithTakeProfitAndStopLoss (LTC/USDT:USDT, limit, buy, 1, 60, 70, 50, [object Object])

{
  info: {
    clOrdId: 'e847386590ce4dBCdb76299fc4b696da',
    ordId: '657619916899565571',
    sCode: '0',
    sMsg: 'Order placed',
    tag: 'e847386590ce4dBC'
  },
  id: '657619916899565571',
  clientOrderId: 'e847386590ce4dBCdb76299fc4b696da',
  symbol: 'LTC/USDT:USDT',
  type: 'limit',
  side: 'buy',
  trades: [],
  reduceOnly: false,
  fees: []
}
```

### CODE REVIEW

File: ts/src/base/Exchange.ts [hunk 1]
```diff
@@ -476,11 +476,15 @@ export default class Exchange {
                 'createMarketOrderWithCost': undefined,
                 'createMarketSellOrderWithCost': undefined,
                 'createOrders': undefined,
+                'createOrderWithTpAndSl': undefined,
```
Comment (carlosmiei) [inline ts/src/base/Exchange.ts:479]: I'm wondering if we should make the name more explicit, like 
createOrderWithStopLossAndTakeProfit

Comment (Dan-krm) [reply ts/src/base/Exchange.ts:479]: `createOrderWithStopLossAndTakeProfit` feels a bit verbose but I think it's a good idea to make them more explicit

File: ts/src/base/Exchange.ts [hunk 2]
```diff
@@ -4135,12 +4139,106 @@ export default class Exchange {
...
+         * @returns {object} an [order structure]{@link https://docs.ccxt.com/#/?id=order-structure}
+         */
+        params['takeProfitPrice'] = takeProfitPrice;
+        if (this.has['createTakeProfitOrder']) {
+            return await this.createOrder (symbol, type, side, amount, price, params);
+        }
+        throw new NotSupported (this.id + ' createTakeProfitOrder() is not supported yet');
+    }
+
+    async createOrderWithTakeProfitAndStopLoss (symbol: string, type: OrderType, side: OrderSide, amount, price = undefined, takeProfit = undefined, stopLoss = undefined, params = {}): Promise<Order> {
```
Comment (carlosmiei) [inline ts/src/base/Exchange.ts:4211]: @Dan-krm Now that I'm thinking, what if the user wants to provide things like stopLossPriceType/amount/etc

should we accept it in params?

Comment (Dan-krm) [reply ts/src/base/Exchange.ts:4211]: That's a good idea I can add more optional parameters to this method, should I add it to the signature like `stopLossPriceType = undefined` or just handle them through the params?

I think just handling them through the params would be best, I'll push some changes tomorrow

Comment (Dan-krm) [reply ts/src/base/Exchange.ts:4211]: I added some extra parameters to `createOrderWithTakeProfitAndStopLoss` let me know if you think we should add or remove any

**Pull request merged**